### PR TITLE
En 9418/wipe freeze nft

### PR DESCRIPTION
**Pull request by sasurobert**
Implemented freeze and wipe for single NFT

### CODE REVIEW

File: vm/systemSmartContracts/esdt.go [hunk 1]
```diff
@@ -612,17 +618,68 @@ func (e *esdt) wipe(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
...
+	token, returnCode := e.basicOwnershipChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	if !token.CanFreeze {
+		e.eei.AddReturnMessage("cannot freeze")
+		return vmcommon.UserError
+	}
+	if string(token.TokenType) == core.FungibleESDT {
+		e.eei.AddReturnMessage("only non fungible tokens can be wiped per nonce")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:639]: changed

File: process/smartContract/builtInFunctions/esdtNFTTransfer.go [hunk 2]
```diff
@@ -304,10 +304,12 @@ func (e *esdtNFTTransfer) addNFTToDestination(
...
-	if !isNew {
-		if currentESDTData.TokenMetaData == nil {
-			return process.ErrWrongNFTOnDestination
-		}
+	err = checkFrozeAndPause(dstAddress, esdtTokenKey, currentESDTData, e.pauseHandler)
+	if err != nil {
+		return err
+	}
+
+	if !isNew && currentESDTData.TokenMetaData != nil {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/builtInFunctions/esdtNFTTransfer.go:312]: It was already an over protection. made it simple.

**Review by SebastianMarian — APPROVED** [review]

File: vm/systemSmartContracts/esdt.go [hunk 3]
```diff
@@ -612,17 +618,68 @@ func (e *esdt) wipe(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
...
+	token, returnCode := e.basicOwnershipChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	if !token.CanFreeze {
+		e.eei.AddReturnMessage("cannot freeze")
+		return vmcommon.UserError
+	}
+	if string(token.TokenType) == core.FungibleESDT {
+		e.eei.AddReturnMessage("only non fungible tokens can be freezed per nonce")
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:639]: maybe remove the `per nonce` ?

File: vm/systemSmartContracts/esdt.go [hunk 4]
```diff
@@ -612,17 +618,68 @@ func (e *esdt) wipe(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
...
+		return vmcommon.UserError
+	}
+
+	if !e.isAddressValid(args.Arguments[2]) {
+		e.eei.AddReturnMessage("invalid address to freeze/unfreeze")
+		return vmcommon.UserError
+	}
+
+	composedArg := append(args.Arguments[0], args.Arguments[1]...)
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:648]: which are the desired arguments? Can you please add a comment above the function signature?

File: vm/systemSmartContracts/esdt.go [hunk 5]
```diff
@@ -631,6 +688,25 @@ func (e *esdt) wipe(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
...
+	if len(args.Arguments) != 3 {
+		e.eei.AddReturnMessage("invalid number of arguments, wanted 3")
+		return vmcommon.FunctionWrongSignature
+	}
+	token, returnCode := e.basicOwnershipChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	if string(token.TokenType) == core.FungibleESDT {
+		e.eei.AddReturnMessage("only non fungible tokens can be wiped per nonce")
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:701]: also here, maybe drop the `per nonce` ?

File: vm/systemSmartContracts/esdt.go [hunk 6]
```diff
@@ -631,6 +688,25 @@ func (e *esdt) wipe(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
...
+		return vmcommon.UserError
+	}
+
+	composedArg := append(args.Arguments[0], args.Arguments[1]...)
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:705]: can you please add above the function's signature the expected arguments?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: process/smartContract/builtInFunctions/esdtNFTTransfer.go [hunk 7]
```diff
@@ -300,14 +300,16 @@ func (e *esdtNFTTransfer) addNFTToDestination(
...
-	currentESDTData, isNew, err := getESDTNFTTokenOnDestination(userAccount, esdtTokenKey, esdtDataToTransfer.TokenMetaData.Nonce, e.marshalizer)
+	currentESDTData, _, err := getESDTNFTTokenOnDestination(userAccount, esdtTokenKey, esdtDataToTransfer.TokenMetaData.Nonce, e.marshalizer)
 	if err != nil && !errors.Is(err, process.ErrNFTTokenDoesNotExist) {
 		return err
 	}
-	if !isNew {
-		if currentESDTData.TokenMetaData == nil {
-			return process.ErrWrongNFTOnDestination
-		}
+	err = checkFrozeAndPause(dstAddress, esdtTokenKey, currentESDTData, e.pauseHandler)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/builtInFunctions/esdtNFTTransfer.go:307]: No need for check

File: vm/systemSmartContracts/esdt.go (hunk 3 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:639]: I would leave it. It is more clear

File: vm/systemSmartContracts/esdt.go [hunk 8]
```diff
@@ -612,17 +618,68 @@ func (e *esdt) wipe(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
 	if returnCode != vmcommon.Ok {
 		return returnCode
 	}
+	returnCode = e.wipeTokenFromAddress(args.Arguments[1], token, args.Arguments[0], args.Arguments[0])
+	return returnCode
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:622]: Would leave as it is

File: vm/systemSmartContracts/esdt.go (hunk 4 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:648]: Is it desired. As this is how we save inside the frozen account

File: vm/systemSmartContracts/esdt.go (hunk 5 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:701]: Would leave as it is

File: vm/systemSmartContracts/esdt.go (hunk 6 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:705]: Will add in next PR.

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**